base/services/w32time/W32Time.c

```diff
@@ -0,0 +1,370 @@
+                        L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\DateTime\\Servers",
+                        0,
+                        KEY_QUERY_VALUE,
+                        &hKey);
+    if (lRet != ERROR_SUCCESS)
+        {
+        return 1;
+        }
+
+    while (TRUE)
+    {
+        dwValSize = MAX_VALUE_NAME * sizeof(WCHAR);
+        szValName[0] = L'\0';
+        lRet = RegEnumValueW(hKey,
+                             dwIndex,
+                             szValName,
+                             &dwValSize,
+                             NULL,
+                             NULL,
+                             (LPBYTE)szData,
+                             &dwNameSize);
+        if (lRet == ERROR_SUCCESS)
+        {
+            /* Get data from default reg value */
+            if (wcscmp(szValName, L"") == 0) // if (Index == 0)
+            {
+                wcscpy(szDefault, szData);
+                dwIndex++;
+            }
+            else
+            {
+                dwIndex++;
+            }
+        }
+        else if (lRet != ERROR_MORE_DATA)
+        {
+            break;
+        }
+    }
+
+
+    dwIndex = 0;
+    while (TRUE)
+    {
+        dwValSize = MAX_VALUE_NAME * sizeof(WCHAR);
+        szValName[0] = L'\0';
+        lRet = RegEnumValueW(hKey,
+                             dwIndex,
+                             szValName,
+                             &dwValSize,
+                             NULL,
+                             NULL,
+                             (LPBYTE)szData,
+                             &dwNameSize);
+        if (lRet == ERROR_SUCCESS)
+        {
+            /* Get date from selected reg value */
+            if (wcscmp(szValName, szDefault) == 0) // if (Index == Default)
+            {
+                wcscpy(szDefault, szData);
+                dwIndex++;
+            }
+            else
+            {
+                dwIndex++;
+            }
+        }
+        else if (lRet != ERROR_MORE_DATA)
+        {
+            break;
+        }
+    }
+
+    RegCloseKey(hKey);
+
+    ulTime = GetServerTime(szDefault);
+
+    if (ulTime != 0)
+        UpdateSystemTime(ulTime);
+
+    return 0;
+}
+
+static VOID CALLBACK
+ServiceMain(int argc, char** argv) 
+{
+    int   error;
+    int   result;
+    DWORD dwPollInterval;
+ 
+    UNREFERENCED_PARAMETER(argc);
+    UNREFERENCED_PARAMETER(argv);
+
+    ServiceStatus.dwServiceType             = SERVICE_WIN32; 
+    ServiceStatus.dwCurrentState            = SERVICE_START_PENDING; 
+    ServiceStatus.dwControlsAccepted        = SERVICE_ACCEPT_STOP | SERVICE_ACCEPT_SHUTDOWN;
+    ServiceStatus.dwWin32ExitCode           = 0; 
+    ServiceStatus.dwServiceSpecificExitCode = 0; 
+    ServiceStatus.dwCheckPoint              = 0; 
+    ServiceStatus.dwWaitHint                = 0; 
+ 
+    hStatus = RegisterServiceCtrlHandler(ServiceName,
+		                         (LPHANDLER_FUNCTION)ControlHandler); 
```

Comment:
Don't cast function pointers. The compiler will correctly complain that the function signature doesn't match -- ControlHandler needs to be WINAPI.

Reply:
I changed ControlHandler to the following:
static DWORD WINAPI ControlHandler(DWORD request) 
and still get compile errors from RosBE without the cast. Suggestions?

Reply:
Should be: `VOID WINAPI ControlHandler(DWORD request);`

Reply:
Yes. That worked. Thanks.